internal/generate/collector/clusterserviceversion.go

```diff
@@ -27,196 +27,218 @@ const (
 	defaultServiceAccountName = "default"
 
 	serviceAccountKind = "ServiceAccount"
+	roleKind           = "Role"
+	cRoleKind          = "ClusterRole"
 )
 
-// SplitCSVPermissionsObjects splits roles that should be written to a CSV as permissions (in)
-// from roles and role bindings that should be written directly to the bundle (out).
-func (c *Manifests) SplitCSVPermissionsObjects() (in, out []client.Object) { //nolint:dupl
-	roleMap := make(map[string]*rbacv1.Role)
-	for i := range c.Roles {
-		roleMap[c.Roles[i].GetName()] = &c.Roles[i]
-	}
-	roleBindingMap := make(map[string]*rbacv1.RoleBinding)
-	for i := range c.RoleBindings {
-		roleBindingMap[c.RoleBindings[i].GetName()] = &c.RoleBindings[i]
-	}
-
-	// Check for unbound roles.
-	for roleName, role := range roleMap {
-		hasRef := false
-		for _, roleBinding := range roleBindingMap {
-			roleRef := roleBinding.RoleRef
-			if roleRef.Kind == "Role" && (roleRef.APIGroup == "" || roleRef.APIGroup == rbacv1.SchemeGroupVersion.Group) {
-				if roleRef.Name == roleName {
-					hasRef = true
-					break
-				}
-			}
-		}
-		if !hasRef {
-			out = append(out, role)
-			delete(roleMap, roleName)
-		}
-	}
-
-	// If a role is bound and:
-	// 1. the binding only has one subject and it is a service account that maps to a deployment service account,
-	//    add the role to in.
-	// 2. the binding only has one subject and it does not map to a deployment service account or is not a service account,
-	//    add both role and binding to out.
-	// 3. the binding has more than one subject and:
-	//    a. one of those subjects is a deployment's service account, add both role and binding to out and role to in.
-	//    b. none of those subjects is a service account or maps to a deployment's service account, add both role and binding to out.
-	deploymentSANames := make(map[string]struct{})
+// SplitCSVPermissionsObjects splits Roles and ClusterRoles bound to ServiceAccounts in Deployments and extraSA's.
+// Roles and ClusterRoles bound to RoleBindings associated with ServiceAccounts are added to inPerms.
+// ClusterRoles bound to ClusterRoleBindings associated with ServiceAccounts are added to inCPerms.
+// All unassociated Roles, ClusterRoles, and bindings are added to out.
+// Any bindings with some associations, but with non-associations, are added to out unmodified.
+func (c *Manifests) SplitCSVPermissionsObjects(extraSAs []string) (inPerms, inCPerms, out []client.Object) { //nolint:gocyclo
```

Comment:
I still want to do some cleanup of this method, and have it accept the full set of ServiceAccounts instead of adding Deployment service accounts internally.